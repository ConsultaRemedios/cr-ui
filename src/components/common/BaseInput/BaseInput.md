### Basic usage
By default, input fill the parent element size
```js
  <BaseInput placeholder="Type something here..."/>
```

### With error
```js
  <BaseInput :has-error="true"/>
```

### Loading State
```js
  <BaseInput :is-loading="true"/>
```

### Disabled State
```js
  <BaseInput :disabled="true"/>
```
### With prefix or suffix icon
```js
  <BaseInput 
    :prefix-icon="{ icon: $getIcon(allIcons, 'place'), color: '#00AAE5' }" 
    :suffix-icon="{ icon: $getIcon(allIcons, 'search'), color: '#666666' }"
    size="small" 
  />
```