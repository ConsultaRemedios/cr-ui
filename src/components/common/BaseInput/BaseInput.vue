<template>
  <div :class="$style.container">
    <BaseIcon
      v-if="prefixIcon"
      :id="prefixIcon.icon.id"
      :class="[$style.inputIcon, $style.prefixIcon]"
      :style="{ color: prefixIcon.color }"
    />

    <input
      :class="inputClasses"
      :name="name"
      :style="customStyle"
      v-bind="$attrs"
      @keydown="onKeydown"
      @input="onChange"
      @blur="onBlur"
      @focus="onFocus"
    >

    <BaseIcon
      v-if="suffixIcon"
      :id="suffixIcon.icon.id"
      :class="[$style.inputIcon, $style.suffixIcon]"
      :style="{ color: suffixIcon.color }"
    />

    <img
      v-if="isLoading"
      :class="$style.loader"
      src="./../../../assets/loader-input.svg"
    >
  </div>
</template>

<script>
import BaseIcon from '../BaseIcon';

export default {

  name: 'BaseInput',

  components: {
    BaseIcon,
  },
  inheritAttrs: false,

  props: {
    classes: {
      type: Array,
      default: () => [],
    },

    size: {
      type: String,
      default: 'medium',
    },
    name: {
      type: String,
      default: '',
    },

    hasError: {
      type: Boolean,
      default: false,
    },

    isLoading: {
      type: Boolean,
      default: false,
    },

    customStyle: {
      type: Object,
      default: () => {},
    },

    model: {
      type: String,
      default: '',
    },

    prefixIcon: {
      type: Object,
    },

    suffixIcon: {
      type: Object,
    },

  },

  computed: {
    inputClasses() {
      return this.classes.concat([this.$style[this.size], {
        [this.$style.input]: true,
        [this.$style.hasError]: this.hasError,
        [this.$style.isLoading]: this.isLoading,
        [this.$style.inputPadLef]: this.prefixIcon,
        [this.$style.inputPadRig]: this.suffixIcon,
      }]);
    },
  },

  methods: {
    createEvent(e) {
      return {
        name: this.model || this.name,
        value: e.currentTarget.value,
        event: e,
      };
    },

    onKeydown(e) {
      this.$emit('keydown', this.createEvent(e));
    },

    onChange(e) {
      this.$emit('change', this.createEvent(e));
    },

    onBlur(e) {
      this.$emit('blur', this.createEvent(e));
    },

    onFocus(e) {
      this.$emit('focus', this.createEvent(e));
    },
  },
};
</script>

<style module>
  .container {
    position: relative;
  }

  .input {
    width: 100%;
    border: 1px solid #DADADA;
    border-radius: 3px;
    color: #333333;
    -webkit-appearance: none;
  }

  .input:-moz-ui-invalid {
    box-shadow: none;
  }

  .small {
    height: 36px;
    padding: 11px 15px;
    font-size: 14px;
  }

  .medium {
    height: 45px;
    padding: 11px 15px;
    font-size: 15px;
  }

  .inputPadLef {
    padding-left: 35px;
  }

  .inputPadRig {
    padding-right: 35px;
  }

  .hasError {
    border-color: #D30B3F;
  }

  .isLoading {
    padding-right: 50px;
  }

  .input:focus {
    border-width: 2px;
    border-color: #00AAE5;
  }

  .input:disabled {
    background-color: #F9F9F9;
    border-color: #DADADA;
  }

  .input::placeholder {
    color: #AAAAAA;
    opacity: 1;
  }

  .loader {
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    right: 18px;
    animation: rotate-loader 1.5s infinite;
  }

  .inputIcon {
    position: absolute;
    top: 50%;
    font-size: 20px;
    transform: translateY(-50%);
  }

  .prefixIcon {
    left: 10px;
  }

  .suffixIcon {
    right: 10px;
  }

  @keyframes rotate-loader {
    0% {
      transform: translateY(-50%) rotate(0deg);
    }

    100% {
      transform: translateY(-50%) rotate(360deg);
    }
  }
</style>
