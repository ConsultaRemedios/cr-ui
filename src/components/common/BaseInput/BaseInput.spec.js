import { mount } from '@vue/test-utils';
import BaseInput from './BaseInput.vue';

describe('BaseInput component matches the snapshot with', () => {
  const defaultProps = {
    name: 'name',
  };
  const defaultAttrs = {
    placeholder: 'placeholder',
  };
  let propsData;

  const mountBaseInput = (props = {}, attributes = {}) => {
    const propsData = { ...defaultProps, ...props };
    const attrs = { ...defaultAttrs, ...attributes };

    return mount(BaseInput, {
      propsData,
      attrs,
    });
  };

  beforeEach(() => {
    propsData = {
      id: 'id',
      name: 'name',
      placeholder: 'placeholder',
      type: 'text',
      size: 'small',
    };
  })

  it('empty props', () => {
    const wrapper = mount(BaseInput, {
      propsData: {}
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('props string', () => {
    const wrapper = mountBaseInput();

    expect(wrapper.element).toMatchSnapshot();
  });

  it('isNumeric true', () => {
    const wrapper = mountBaseInput({}, {
      type: 'tel',
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('isLoading true', () => {
    const wrapper = mountBaseInput({ isLoading: true });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('disabled true', () => {
    const wrapper = mountBaseInput(null, { disabled: true });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('readOnly true', () => {
    const wrapper = mountBaseInput(null, { readonly: true });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('autofocus true', () => {

    const wrapper = mountBaseInput(null, {
      autofocus: 'autofocus'
    });

    expect(wrapper.find('input').attributes().autofocus).toEqual('autofocus');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('matches snapshot when there is a prefix icon', () => {
    propsData.prefixIcon = {
      icon: {},
      color: '#666',
    };

    const wrapper = mount(BaseInput, {
      propsData,
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('matches snapshot when there is a suffix icon', () => {
    propsData.suffixIcon = {
      icon: {},
      color: '#666',
    };

    const wrapper = mount(BaseInput, {
      propsData,
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  describe('with one custom class', () => {
    it('adds custom class on input', () => {
      propsData.classes = ['the-custom-class'];
      const wrapper = mount(BaseInput, { propsData });

      expect(wrapper.find('input').classes()).toContain('the-custom-class')
    });
  });

  describe('with more than one custom classes', () => {
    it('adds custom classes on input', () => {
      propsData.classes = ['the-custom-class', 'is-custom', 'really'];
      const wrapper = mount(BaseInput, { propsData });

      expect(wrapper.find('input').classes()).toContain('the-custom-class');
      expect(wrapper.find('input').classes()).toContain('is-custom');
      expect(wrapper.find('input').classes()).toContain('really');
    });
  });

  describe('Behavior', () => {
    describe('When input event is emitted by input', () => {
      describe('and model prop is empty', () => {
        const wrapper = mountBaseInput({ model: 'name', name: 'user[name]' }, { value: 'John Doe' });

        it('emits change event with the proper params', () => {
          jest.spyOn(wrapper.vm, '$emit');
          wrapper.find('input').trigger('input');

          const emitted = wrapper.vm.$emit.mock.calls[0];

          expect(emitted[1]).toMatchObject({ name: 'name', value: 'John Doe' });
          expect(wrapper.vm.$emit).toHaveBeenCalledTimes(1);
          expect(emitted[0]).toEqual('change');
          expect(emitted[1].event.target).toEqual(wrapper.find('input').element);
        });
      });

      describe('and model prop is not empty', () => {
        const wrapper = mountBaseInput({  name: 'user[name]' }, { value: 'John Doe' });

        it('emits change event with the proper params', () => {
          jest.spyOn(wrapper.vm, '$emit');
          wrapper.find('input').trigger('input');

          const emitted = wrapper.vm.$emit.mock.calls[0];

          expect(emitted[0]).toEqual('change');
          expect(emitted[1]).toMatchObject({ name: 'user[name]', value: 'John Doe' });
          expect(emitted[1].event.target).toEqual(wrapper.find('input').element);
        });
      });
    });

    describe('When blur event is emitted by input', () => {
      describe('and model prop is empty', () => {
        const wrapper = mountBaseInput({ name: 'user[name]', model: 'name' }, { value: 'John Doe' });

        it('emits "blur" event with the proper params', () => {
          jest.spyOn(wrapper.vm, '$emit');
          wrapper.find('input').trigger('blur');

          const emitted = wrapper.vm.$emit.mock.calls[0];

          expect(wrapper.vm.$emit).toHaveBeenCalledTimes(1);
          expect(emitted[0]).toEqual('blur');
          expect(emitted[1]).toMatchObject({ name: 'name', value: 'John Doe' });
          expect(emitted[1].event.target).toEqual(wrapper.find('input').element);
        });
      });

      describe('and model prop is not empty', () => {
        const wrapper = mountBaseInput({ name: 'user[name]' }, { value: 'John Doe' });

        it('emits "blur" event with the proper params', () => {
          jest.spyOn(wrapper.vm, '$emit');
          wrapper.find('input').trigger('blur');

          const emitted = wrapper.vm.$emit.mock.calls[0];

          expect(emitted[0]).toEqual('blur');
          expect(emitted[1].name).toEqual('user[name]')
          expect(emitted[1].value).toEqual('John Doe')
          expect(emitted[1].event.target).toEqual(wrapper.find('input').element);
        });
      });
    });

    it('emits blur event with proper params when blur event is emitter on input', () => {
      const wrapper = mountBaseInput({ name: 'email' }, { value: 'john@doe.com' });
      jest.spyOn(wrapper.vm, '$emit');

      expect(wrapper.vm.$emit).not.toHaveBeenCalled();

      wrapper.find('input').trigger('blur');

      expect(wrapper.vm.$emit).toHaveBeenCalledTimes(1);

      const emitted = wrapper.vm.$emit.mock.calls[0];

      expect(emitted[0]).toEqual('blur');
      expect(emitted[1].name).toEqual('email');
      expect(emitted[1].value).toEqual('john@doe.com');
      expect(emitted[1].event.target).toMatchObject(wrapper.find('input').element);
    });
  });

  describe('Render', () => {
    describe('When there is id', () => {
      it('renders the snapshot', () => {
        const wrapper = mountBaseInput(null, {
          id: 'input-term'
        });

        expect(wrapper.element).toMatchSnapshot();
        expect(wrapper.find('input').attributes().id).toEqual('input-term');
      });
    });

    describe('When there is no id', () => {
      it('renders the snapshot', () => {
        const wrapper = mountBaseInput({ size: 'small' });
        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });

  describe('Event Listeners', () => {
    describe('When keydown event happens', () => {
      it('calls method onKeydown', () => {
        const wrapper = mountBaseInput(null, { value: 'abc' });
        const input = wrapper.find('input');

        jest.spyOn(wrapper.vm, '$emit');

        input.trigger('keydown');
        expect(input.element.value).toBe('abc');
        expect(wrapper.vm.$emit).toHaveBeenCalled();
        expect(wrapper.vm.$emit.mock.calls[0][0]).toEqual('keydown');
        expect(wrapper.vm.$emit.mock.calls[0][1].value).toEqual('abc');
      });
    });

    describe('When changed', () => {
      let wrapper;
      let input;

      beforeAll(() => {
        wrapper = mountBaseInput();
        input = wrapper.find('input');
        jest.spyOn(wrapper.vm, '$emit');
      });

      it('calls method "onChange"', () => {
        input.trigger('input');

        expect(wrapper.vm.$emit).toHaveBeenCalled();
        expect(wrapper.vm.$emit.mock.calls[0][0]).toEqual('change');
      });
    });

    describe('When blur', () => {
      let wrapper;
      let input;

      beforeAll(() => {
        wrapper = mountBaseInput();
        input = wrapper.find('input');
        jest.spyOn(wrapper.vm, '$emit');
      });

      it('calls method "onBlur"', () => {
        input.trigger('blur');

        expect(wrapper.vm.$emit).toHaveBeenCalled();
        expect(wrapper.vm.$emit.mock.calls[0][0]).toEqual('blur');
      });
    });

    describe('When focus', () => {
      let wrapper;
      let input;

      beforeAll(() => {
        wrapper = mountBaseInput();
        input = wrapper.find('input');
        jest.spyOn(wrapper.vm, '$emit');
      });

      it('calls method "onFocus"', () => {
        input.trigger('focus');

        expect(wrapper.vm.$emit).toHaveBeenCalled();
        expect(wrapper.vm.$emit.mock.calls[0][0]).toEqual('focus');
      });
    });
  });

  describe('Methods', () => {
    const event = {
      currentTarget: {
        value: 'value',
        isTrusted: true,
      },
    };
    const formatedEvent = {
      name: 'name',
      value: 'value',
      event: {
        currentTarget: {
          value: 'value',
          isTrusted: true,
        },
      }
    };

    describe('#createEvent', () => {
      describe('When passing a event', () => {
        const wrapper = mountBaseInput(null, { name: 'name' });

        it('parses event in a new object', () => {
          expect(wrapper.vm.createEvent(event)).toEqual(formatedEvent);
        });
      });
    });

    describe('#onKeydown', () => {
      it('emits "keydown" event with proper payload', () => {
        const wrapper = mountBaseInput(null, { name: 'name' });
        jest.spyOn(wrapper.vm, '$emit');

        expect(wrapper.vm.$emit).not.toHaveBeenCalled();

        wrapper.vm.onKeydown(event)
        expect(wrapper.vm.$emit).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.$emit).toHaveBeenCalledWith('keydown', formatedEvent);
      });
    });

    describe('#onChange', () => {
      describe('When called', () => {
        let wrapper;

        beforeEach(() => {
          wrapper = mountBaseInput(null, { name: 'name' });
          jest.spyOn(wrapper.vm, '$emit');
          wrapper.vm.onChange(event)
        });

        it('emits "change" event', () => {
          expect(wrapper.vm.$emit).toHaveBeenCalledWith(
            'change',
            formatedEvent
          );
        });
      });
    });

    describe('#onBlur', () => {
      describe('When called', () => {
        let wrapper;

        beforeEach(() => {
          wrapper = mountBaseInput(null, { name: 'name' });
          jest.spyOn(wrapper.vm, '$emit');
          wrapper.vm.onBlur(event)
        });

        it('emits "blur" event', () => {
          expect(wrapper.vm.$emit).toHaveBeenCalledWith(
            'blur',
            formatedEvent
          );
        });
      });
    });

    describe('#onFocus', () => {
      describe('When called', () => {
        let wrapper;

        beforeEach(() => {
          wrapper = mountBaseInput(null, { name: 'name' });
          jest.spyOn(wrapper.vm, '$emit');
          wrapper.vm.onFocus(event)
        });

        it('emits "focus" event', () => {
          expect(wrapper.vm.$emit).toHaveBeenCalledWith(
            'focus',
            formatedEvent
          );
        });
      });
    });
  });
});
