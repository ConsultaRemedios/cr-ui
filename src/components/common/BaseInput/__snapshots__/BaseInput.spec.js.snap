// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseInput component matches the snapshot with Render When there is id renders the snapshot 1`] = `
<div
  class="container"
>
  <!---->
   
  <input
    class="medium input"
    id="input-term"
    name="name"
    placeholder="placeholder"
  />
   
  <!---->
   
  <!---->
</div>
`;

exports[`BaseInput component matches the snapshot with Render When there is no id renders the snapshot 1`] = `
<div
  class="container"
>
  <!---->
   
  <input
    class="small input"
    name="name"
    placeholder="placeholder"
  />
   
  <!---->
   
  <!---->
</div>
`;

exports[`BaseInput component matches the snapshot with autofocus true 1`] = `
<div
  class="container"
>
  <!---->
   
  <input
    autofocus="autofocus"
    class="medium input"
    name="name"
    placeholder="placeholder"
  />
   
  <!---->
   
  <!---->
</div>
`;

exports[`BaseInput component matches the snapshot with disabled true 1`] = `
<div
  class="container"
>
  <!---->
   
  <input
    class="medium input"
    disabled="disabled"
    name="name"
    placeholder="placeholder"
  />
   
  <!---->
   
  <!---->
</div>
`;

exports[`BaseInput component matches the snapshot with empty props 1`] = `
<div
  class="container"
>
  <!---->
   
  <input
    class="medium input"
    name=""
  />
   
  <!---->
   
  <!---->
</div>
`;

exports[`BaseInput component matches the snapshot with isLoading true 1`] = `
<div
  class="container"
>
  <!---->
   
  <input
    class="medium input isLoading"
    name="name"
    placeholder="placeholder"
  />
   
  <!---->
   
  <img
    class="loader"
    src="./../../../assets/loader-input.svg"
  />
</div>
`;

exports[`BaseInput component matches the snapshot with isNumeric true 1`] = `
<div
  class="container"
>
  <!---->
   
  <input
    class="medium input"
    name="name"
    placeholder="placeholder"
    type="tel"
  />
   
  <!---->
   
  <!---->
</div>
`;

exports[`BaseInput component matches the snapshot with matches snapshot when there is a prefix icon 1`] = `
<div
  class="container"
>
  <svg
    class="icon inputIcon prefixIcon"
    style="color: rgb(102, 102, 102);"
    viewBox="0 0 24 24"
  >
    <use
      xlink:href="#undefined"
    />
  </svg>
   
  <input
    class="small input inputPadLef"
    id="id"
    name="name"
    placeholder="placeholder"
    type="text"
  />
   
  <!---->
   
  <!---->
</div>
`;

exports[`BaseInput component matches the snapshot with matches snapshot when there is a suffix icon 1`] = `
<div
  class="container"
>
  <!---->
   
  <input
    class="small input inputPadRig"
    id="id"
    name="name"
    placeholder="placeholder"
    type="text"
  />
   
  <svg
    class="icon inputIcon suffixIcon"
    style="color: rgb(102, 102, 102);"
    viewBox="0 0 24 24"
  >
    <use
      xlink:href="#undefined"
    />
  </svg>
   
  <!---->
</div>
`;

exports[`BaseInput component matches the snapshot with props string 1`] = `
<div
  class="container"
>
  <!---->
   
  <input
    class="medium input"
    name="name"
    placeholder="placeholder"
  />
   
  <!---->
   
  <!---->
</div>
`;

exports[`BaseInput component matches the snapshot with readOnly true 1`] = `
<div
  class="container"
>
  <!---->
   
  <input
    class="medium input"
    name="name"
    placeholder="placeholder"
    readonly="readonly"
  />
   
  <!---->
   
  <!---->
</div>
`;
